docs/typescript.md

```diff
@@ -1,14 +1,644 @@
+- Often, the compiler will tell us exactly what the target type for an assertion needs to be, enabling us to avoid `as any`.
+
+###### Example (2ee8f56a-e3be-417b-a2c0-260c1319b755)
+
+```typescript
+// Error: Argument of type '"getNftInformation"' is not assignable to parameter of type 'keyof NftController'.ts(2345)
+// 'getNftInformation' is a private method of class 'NftController'
+sinon.stub(nftController, 'getNftInformation');
+```
+
+🚫 `as any`
+
+```typescript
+sinon.stub(nftController, 'getNftInformation' as any);
+```
+
+✅ Compiler specifies that the target type should be `keyof NftController`
+
+```typescript
+sinon.stub(nftController, 'getNftInformation' as keyof typeof nftController);
+```
+
+- Even an assertion to a wrong type still allows the compiler to show us warnings and errors as the code changes, and is therefore preferrable to the dangerous radio silence enforced by `any`.
+- For type assertions to an incompatible shape, use `as unknown as` as a last resort rather than `any` or `as any`.
+
+##### `as` is always acceptable to use for TypeScript syntax other than type assertion
+
+- Writing type guards often requires using the `as` keyword.
+
+###### Example (f16df571-266e-4030-b002-49554558ccd7)
+
+```typescript
+function isFish(pet: Fish | Bird): pet is Fish {
+  return (pet as Fish).swim !== undefined;
+}
+```
+
+- Key remapping in mapped types uses the `as` keyword.
+
+###### Example (6ffd8c99-4768-42e1-8cb7-5710d14f8552)
+
+```typescript
+type MappedTypeWithNewProperties<Type> = {
+  [Properties in keyof Type as NewKeyType]: Type[Properties];
+};
+```
+
+##### `as` is acceptable to use for typing data objects whose shape and contents are determined at runtime, externally, or through deserialization
+
+Preferably, this typing should be accompanied by runtime schema validation performed with type guards and unit tests.
+
+- e.g. The output of `JSON.parse()` or `await response.json()` for a known JSON input.
+- e.g. The type of a JSON file.
+
+<!-- TODO: Add example -->
+
+##### `as` may be acceptable to use in tests, for mocking or to exclude irrelevant but required properties from an input object
+
+It's recommended to provide accurate typing if there's any chance that omitting properties affects the accuracy of the test.
+
+If that is not the case, however, mocking only the properties needed in the test improves readability. It makes the intention and scope of the mock clear, and is more convenient to write, which can encourage test writing and improve test coverage.
+
+<!-- TODO: Add examples -->
+
+### Compiler Directives
+
+TypeScript provides several directive comments that can be used to suppress TypeScript compiler errors. Using these to ignore typing issues is dangerous and reduces the overall effectiveness of TypeScript.
+
+Of these, we will discuss the use cases and pitfalls of `@ts-expect-error` and `any`.
+
+`@ts-ignore`, `@ts-nocheck`, `@ts-check` are disabled by ESLint rules.
```

Comment:
Do we want to link to our `@metamask/eslint-config-typescript` package where this is disabled? Perhaps not the exact line in case it ever changes, but just the package itself: https://github.com/MetaMask/eslint-config/tree/main/packages/typescript

Reply:
The new changes go into more detail, so I'm good on this.